archive/multisource.go

```diff
@@ -3,81 +3,76 @@ package archive
 import (
 	"context"
 	"io"
+	"sync/atomic"
 
-	"github.com/brimsec/zq/ast"
 	"github.com/brimsec/zq/driver"
 	"github.com/brimsec/zq/field"
-	"github.com/brimsec/zq/filter"
 	"github.com/brimsec/zq/pkg/iosrc"
 	"github.com/brimsec/zq/scanner"
 	"github.com/brimsec/zq/zbuf"
 	"github.com/brimsec/zq/zio"
 	"github.com/brimsec/zq/zio/detector"
 	"github.com/brimsec/zq/zio/zngio"
 	"github.com/brimsec/zq/zng/resolver"
+	"go.uber.org/multierr"
 )
 
-type multiCloser struct {
-	closers []io.Closer
+type Source interface {
+	driver.MultiSource
+	Stats() ChunkStats
 }
 
-func (c *multiCloser) Close() error {
-	var err error
-	for _, c := range c.closers {
-		if closeErr := c.Close(); err == nil {
-			err = closeErr
+type multiCloser []io.Closer
+
+func (c multiCloser) Close() (err error) {
+	for _, closer := range c {
+		if closeErr := closer.Close(); closeErr != nil {
+			err = multierr.Append(err, closeErr)
 		}
 	}
-	return err
+	return
 }
 
 type scannerCloser struct {
 	scanner.Scanner
 	io.Closer
 }
 
-func newSpanScanner(ctx context.Context, ark *Archive, zctx *resolver.Context, f filter.Filter, filterExpr ast.BooleanExpr, si SpanInfo) (sc *scannerCloser, err error) {
-	if len(si.Chunks) == 1 {
-		rc, err := iosrc.NewReader(ctx, si.Chunks[0].Path(ark))
-		if err != nil {
-			return nil, err
-		}
-		sn, err := scanner.NewScanner(ctx, zngio.NewReader(rc, zctx), f, filterExpr, si.Span)
-		if err != nil {
-			rc.Close()
-			return nil, err
-		}
-		return &scannerCloser{sn, rc}, nil
-	}
-	closers := make([]io.Closer, 0, len(si.Chunks))
-	defer func() {
-		if err != nil {
-			for _, c := range closers {
-				c.Close()
-			}
-		}
-	}()
+func newSpanScanner(ctx context.Context, ark *Archive, zctx *resolver.Context, sf driver.SourceFilter, si SpanInfo) (sc *scannerCloser, stats ChunkStats, err error) {
+	closers := make(multiCloser, 0, len(si.Chunks))
 	readers := make([]zbuf.Reader, 0, len(si.Chunks))
 	for _, chunk := range si.Chunks {
-		rc, err := iosrc.NewReader(ctx, chunk.Path(ark))
+		rc, err := newChunkReader(ctx, chunk, ark, sf.Span)
```

Comment:
This should use the span from the SpanInfo, not the query filter. Would you make sure there's a test whose query span is (0, max) so that you can verify this?

```suggestion
		rc, err := newChunkReader(ctx, chunk, ark, si.Span)
```

Reply:
@alfred-landrum I'm not sure what I would be testing for. Don't most of our tests already assume something similar to [0, max]?

The first few lines of `newChunkReader` intersect the chunk's span with the query (or spanInfo) span: 
```
func newChunkReader(ctx context.Context, chunk Chunk, ark *Archive, span nano.Span) (*chunkReader, error) {
	cspan := chunk.Span()
	span = cspan.Intersect(span)
```

Reply:
The reason to use the SpanInfo span is because it describes the just portion of the chunk needed for the scanner instance returned from newSpanScanner. Using a dataset with overlaps highlights the difference, here's an example:

```
$ zar import -s 128MiB ~/work/data/wrccdc-year1/zeek-logs.zng
# using sf.Span
$ zar zq -t -s "count()"
#0:record[count:uint64]
0:[96217188;]
data opened: 19038086607
data read:   19038086607

# using si.Span
$ zar zq -t -s "count()"
#0:record[count:uint64]
0:[96217188;]
data opened: 19038086607
data read:   4833249534
```

Regarding tests: to have a test that verifies this, it's easier with a dataset containing overlaps, and shouldn't need to use a query span.

Reply:
Gotcha. This is because the same chunk can be used in different SpanInfos. Will fix.